aggregator/elem.go

```diff
@@ -192,17 +194,27 @@ func (e *CounterElem) Consume(earlierThanNanos int64, fn aggMetricFn) bool {
 		if e.values[idx].timeNanos >= earlierThanNanos {
 			break
 		}
-		endAtNanos := e.values[idx].timeNanos + int64(e.policy.Resolution().Window)
-		e.processValue(endAtNanos, e.values[idx].counter, fn)
 		idx++
 	}
+	e.toConsume = e.toConsume[:0]
 	if idx > 0 {
 		// Shift remaining values to the left and shrink the values slice
+		e.toConsume = append(e.toConsume, e.values[:idx]...)
 		n := copy(e.values[0:], e.values[idx:])
 		e.values = e.values[:n]
 	}
 	canCollect := len(e.values) == 0 && e.tombstoned
 	e.Unlock()
+
+	// Process the counters.
+	for i := range e.toConsume {
+		endAtNanos := e.toConsume[i].timeNanos + int64(e.policy.Resolution().Window)
+		e.toConsume[i].counter.Lock()
+		e.processValue(endAtNanos, e.toConsume[i].counter.Counter, fn)
```

Comment:
Do we need to lock the counter when we call `processValue`? Just thinking that a possible alternative would be to use an atomic int64 and do atomic loads and increments. It seems `processValue` just calls  `fn` with the counter's current value so if we're okay with allowing updates to the counter after we load its value we won't need a lock.